File: 0826/solution.java
import java.util.Arrays;

class Solution {
    public int maxProfitAssignment(int[] difficulty, int[] profit, int[] workers) {
        int[][] pairs = new int[profit.length][2];
        for (int i = 0; i < profit.length; i++) {
            pairs[i][0] = difficulty[i];
            pairs[i][1] = profit[i];
        }
        Arrays.sort(pairs, (a, b) -> Integer.compare(a[0], b[0]));
        Arrays.sort(workers);
        int p = 0;
        int bestProfit = 0;
        int totalProfit = 0;
        for (int w : workers) {
            while (p < pairs.length && w >= pairs[p][0]) {
                bestProfit = Math.max(bestProfit, pairs[p++][1]);
            }
            totalProfit += bestProfit;
        }
        return totalProfit;
    }
}

// time O(nlogn + mlogm)
// space O(n)

File: 0826/solution.ts
function maxProfitAssignment(
  difficulty: number[],
  profit: number[],
  workers: number[]
): number {
  const pairs: Array<[number, number]> = [];

  for (let i = 0; i < profit.length; i++) {
    pairs.push([difficulty[i], profit[i]]);
  }
  pairs.sort((a, b) => a[0] - b[0]);
  workers.sort((a, b) => a - b);
  let p = 0;
  let bestProfit = 0;
  let totalProfit = 0;
  for (let w of workers) {
    while (p < pairs.length && w >= pairs[p][0]) {
      bestProfit = Math.max(bestProfit, pairs[p++][1]);
    }
    totalProfit += bestProfit;
  }
  return totalProfit;
}
